api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto

```diff
@@ -261,6 +262,20 @@ message ExtAuthz {
   // It's recommended you set this to true unless you already rely on the old behavior. False is the
   // default only for backwards compatibility.
   bool encode_raw_headers = 23;
+
+  // Rules for what modifications an ext_authz server may make to the request headers before
+  // continuing decoding / forwarding upstream.
+  //
+  // In ext_authz, some of the default values for this proto are overridden for backward
+  // compatibility. The following fields' default values are overridden (both set to true):
+  // :ref:`allow_all_routing <envoy_v3_api_field_config.common.mutation_rules.v3.HeaderMutationRules.allow_all_routing>`
+  // :ref:`allow_envoy <envoy_v3_api_field_config.common.mutation_rules.v3.HeaderMutationRules.allow_envoy>`
+  //
+  // The above has no effect if the fields are set explicitly in your config.
+  //
+  // Note that no matter what rules are set here, ext_authz can never remove or append
+  // pseudo headers.
+  config.common.mutation_rules.v3.HeaderMutationRules decoder_header_mutation_rules = 26;
```

Comment:
I have checked you implementation in more detail. Here are some comments:
1. Rather than override the default value and document it here, it would be better to use optional header mutation checker. I mean this field is not set explicitly, just ignore it and don't do the mutation check.
2. What the relationship between previous `validate_mutations` field (it was added at 05/09, seems we have introduced a tech debt) and this field? Seems this field provide more powerful feature?

Reply:
1.

I initially did what you are describing. I don't like it, because HeaderMutationRules can then configure behavior changes that you did not explicitly ask for.

E.g. Imagine I just set `decoder_header_mutation_rules.disallow_expression` to hide some internal header from the ext_authz service. Well, now the rest of HeaderMutationRules defaults apply and there are behavior changes unrelated to what I actually configured.

Which would mean here we would instead have to document "if you configure this field, make sure you also configure these sub fields lest they silently change the filter behavior", which doesn't sound better to me than what I have done here.

2.

The fields do different things.

`validate_mutations` checks for "correctness" of header mutations and query parameters added by the ext_authz service before adding them to the header map / path.

`decoder_header_mutation_rules` allows the filter to reject mutations to specific headers / classes of headers.

Reply:
To 1:
From the perspective of the API, even no one field of `HeaderMutationRules` is set and only a empty config is provided by `decoder_header_mutation_rules: {}`. It's still an explicitly setting. The default value that recorded in the document of HeaderMutationRules message self should be applied.

It's not a good idea to change the default value of message at the position where the message is embedded, except there is a very strong point. Maybe the default value of HeaderMutationRules doesn't meet you expectation, but it actually doesn't break the backward compatibility.

To 2:

Thanks. It make sense. Could you also add a simple explanation to the comment of the new field incase someone may have similar confusion about that.

Reply:
Thanks for explaining. That all makes sense to me. Done and done.